### PR TITLE
Script to build a signed .phar for drag-n-drop deployment.

### PR DESCRIPTION
**Pull request by paragonie-scott**
Supersedes #130 -- adds optional OpenSSL signing

### CODE REVIEW

**Comment by paragonie-scott** [comment]
Testing instructions:

```sh
openssl genrsa -out private.pem 2048
./build-phar.sh private.pem
```

You should see a .phar and a .phar.pubkey file, which both need to be present.

**Comment by paragonie-scott** [comment]
Looks like phpunit is breaking Travis. (This PR entails no change to code covered by the unit tests.)

**Comment by paragonie-scott** [comment]
Don't merge yet, getting:

```
PHP Fatal error:  Uncaught exception 'PharException' with message 'unable to copy stub of old phar to new phar "/var/www/defuse/php-encryption/dist/defuse-crypto.phar"' in /var/www/defuse/php-encryption/other/build_phar.php:37
Stack trace:
#0 /var/www/defuse/php-encryption/other/build_phar.php(37): Phar->setSignatureAlgorithm(16, '-----BEGIN PRIV...')
#1 {main}
  thrown in /var/www/defuse/php-encryption/other/build_phar.php on line 37
```

**Comment by paragonie-scott** [comment]
test/keyed.php
```php
<?php
use \Defuse\Crypto\Crypto;

$phar = new \Phar(dirname(__DIR__).'/dist/defuse-crypto.phar');
$signature = $phar->getSignature();
var_dump($signature);

require_once dirname(__DIR__).'/dist/defuse-crypto.phar';

$key = Crypto::createNewRandomKey();
$data = Crypto::encrypt('test message', $key);
var_dump(
    $data,
    Crypto::decrypt($data, $key)
);
```

Output:

```output
array(2) {
  ["hash"]=>
  string(512) "9F735E596E59C398F4FF34DCA350179FE6E8590608133F89A32E5F6EF822DBF255FF40247771702CD97FBC7354C462A6E3267511DF8EB2C3FEEB09A1B77F14B5EF70903C133BA5AC51C6BB29C87472981E218F611C3D61E4F969B7B0F1AFB4225FD02C93CE38C78FCE13E57C01E054AEEF2628E1E044799C765CE7DED29B4A5AC49A27C917C0F40FE45C42525FF83EFBE58468F412847F0A3C87CDA922FF21373BB76ABB439E0124532E699C5C293A0D4F97378F8A669EAF21B2B2B0F567A9FE9BC273C046664F9A1C6FEB6934D079F018DE3A4AF7291DC060D50DC77FC7535ED0EA61CA2633201D53DAC7B14AB958E319F6DD9969D277D366FE1D87F9939F9B"
  ["hash_type"]=>
  string(7) "OpenSSL"
}
string(160) "def50200c5754eb2c9f5fbbdf6b070cb269e2955339e5a36e4fede1b8e43c32dac5bcba5cf7765d3a94d3ac984ef11e4109843d999d4e9b1a1fc3aa9ef34a1933eb967dcef95afefcc189f0bbfbcb8dc"
string(12) "test message"
```

Yay, it works.

**Comment by paragonie-scott** [comment]
After merging, @defuse please generate an OpenSSL key pair and publish your public key in `dist/defuse-crypto.phar.pubkey` and GPG sign it. (Upon verifying it, I will publish my own signature with my long-term GPG key.)

Some questions I still have (but have yet investigated):

1. Does this work with NIST P256 keypairs, or are we stuck with RSA forever?
2. Does Phar support RSASS-PSS + MGF1-SHA256 or does everything use PKCS1v1.5 (the OpenSSL default)?
3. Can this bad/annoying Phar behavior be fixed?

**Comment by defuse** [comment]
Ok, might take me a while. I want to do it on my airgap.

**Comment by paragonie-scott** [comment]
So, I was thinking about doing something like this, in the end:

```dist/
|||\
||\ `- README.md
|\ `-- defuse-crypto.phar
\ `--- defuse-crypto.phar.pubkey
 `---- pubkey-signatures/
       |||\
	   ||\ `- defuse.pgp.sig
	   |\ `-- paragonie.pgp.sig
	   \ `--- other-community-member-1.pgp.sig
	    `---- other-community-member-2.pgp.sig
```

Each entry in `pubkey-signatures/` would include a PGP signature of `defuse-crypto.phar.pubkey`  for various members of the PHP community to attest for the authenticity of the public key.

(This wouldn't add any sort of automated verification to the public key, but even if someone breaches you they'd have to breach, at the minimum, me as well before they could trick someone into running a trojan'd copy of this library.)

File: src/index.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            ['\\', '_'],
+            '/',
+            $relative_class
+        ).'.php';
+
+    // If the file exists, require it
+    if (\file_exists($file)) {
+        require $file;
+    }
+});
```
Comment (defuse) [inline src/index.php]: This needs to be updated with the changes to `autoload.php`. Also could we somehow avoid putting this in revision control and confusing people, like putting it in `other/` and temporarily copying *only* when we're building the phar?

Comment (paragonie-scott) [inline src/index.php]:
```
mv autoload.php index.php
./do_stuff
mv index.php autoload.php
```

...seems like the cleanest workaround.

Comment (defuse) [inline src/index.php]: +1

**Comment by paragonie-scott** [comment]
This should be good to go.

**Comment by defuse** [comment]
LGTM.

**Pull request merged**